packages/backend-common/src/reading/GerritUrlReader.ts

```diff
@@ -109,8 +127,60 @@ export class GerritUrlReader implements UrlReader {
     );
   }
 
-  async readTree(): Promise<ReadTreeResponse> {
-    throw new Error('GerritReader does not implement readTree');
+  async readTree(
+    url: string,
+    options?: ReadTreeOptions,
+  ): Promise<ReadTreeResponse> {
+    const { filePath } = parseGitilesUrl(this.integration.config, url);
+    const apiUrl = getGerritBranchApiUrl(this.integration.config, url);
+    let response: Response;
+    try {
+      response = await fetch(apiUrl, {
+        method: 'GET',
+        ...getGerritRequestOptions(this.integration.config),
+      });
+    } catch (e) {
+      throw new Error(`Unable to read branch state ${url}, ${e}`);
+    }
+
+    if (response.status === 404) {
+      throw new NotFoundError(`Not found: ${url}`);
+    }
+
+    if (!response.ok) {
+      throw new Error(
+        `${url} could not be read as ${apiUrl}, ${response.status} ${response.statusText}`,
+      );
+    }
+    const branchInfo = (await parseGerritJsonResponse(response as any)) as {
+      revision: string;
+    };
+    if (options?.etag === branchInfo.revision) {
+      throw new NotModifiedError();
+    }
+
+    const git = Git.fromAuth({
+      username: this.integration.config.username,
+      password: this.integration.config.password,
+    });
+    const tempDir = await createTemporaryDirectory();
+    // The "fromTarArchive" function will strip the top level directory so
+    // an additional directory level is created when we clone.
+    await git.clone({
+      url: getGerritCloneRepoUrl(this.integration.config, url),
```

Comment:
Thinking it's best if we use `branchInfo.revision` when we clone here, just so we don't end up cloning something that doesn't match the returned `etag`.